// @antv/l7-utils/lib/stencli.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/stencli.ts
var stencli_exports = {};
__export(stencli_exports, {
  getMask: () => getMask
});
module.exports = __toCommonJS(stencli_exports);
function getMask(mask, maskInside) {
  return {
    enable: mask,
    mask: 255,
    func: {
      cmp: 514,
      ref: maskInside ? 1 : 0,
      mask: 255
    }
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getMask
});
